api/src/main/java/org/pmiops/workbench/actionaudit/auditors/SumoLogicAuditorImpl.kt

```diff
@@ -0,0 +1,164 @@
+                dbWorkspace.firecloudName)
+        val vmOwner = userRoles
+                .map { userDao.findUserByUsername(it.email) }
+                .filter { dbUserToVmName(it).equals(event.vmName) }
+                .firstOrNull()
+
+        var agentEmail = ""
+        var agentId = 0L
+        if (vmOwner != null) {
+            agentEmail = vmOwner.username
+            agentId = vmOwner.userId
+        } else {
+            // If the VM name doesn't match a user on the workspace, we'll still log an
+            // event in the target workspace, but with nulled-out user info.
+            logger.warning(String.format("Could not find a user for VM name %s in project %s", event.vmName, event.projectName))
+        }
+
+        val actionId = actionIdProvider.get()
+        fireEventSet(ActionAuditEvent(
+                timestamp = clock.millis(),
+                actionId = actionId,
+                actionType = ActionType.HIGH_EGRESS,
+                agentType = AgentType.USER,
+                agentId = agentId,
+                agentEmailMaybe = agentEmail,
+                targetType = TargetType.WORKSPACE,
+                targetIdMaybe = dbWorkspace.workspaceId
+        ), event, null)
+    }
+
+    override fun fireFailedToParseEgressEvent(request: EgressEventRequest) {
+        fireEventSet(getGenericBaseEvent(), null, String.format(
+                "Failed to parse egress event JSON from SumoLogic. Field contents: %s",
+                request.eventsJsonArray))
+    }
+
+    override fun fireBadApiKeyEgressEvent(apiKey: String, request: EgressEventRequest) {
+        fireEventSet(getGenericBaseEvent(), null, String.format(
+                "Received bad API key from SumoLogic. Bad key: %s, full request: %s",
+                apiKey, request.toString()))
+    }
+
+    private fun fireFailedToFindWorkspace(event: EgressEvent) {
+        fireEventSet(getGenericBaseEvent(), event, String.format(
+                "Failed to find workspace for high-egress event: %s",
+                event.toString()))
+    }
+
+    /**
+     * Creates and fires a set of events derived from the given base event, incorporating audit rows
+     * to record properties of the egress event and a human-readable comment. Either the egress event
+     * or the comment may be null, in which case those row(s) won't be generated.
+     */
+    private fun fireEventSet(baseEvent: ActionAuditEvent, egressEvent: EgressEvent?, comment: String?) {
+        var events = ArrayList<ActionAuditEvent>()
+        if (egressEvent != null) {
+            val propertyValues = TargetPropertyExtractor.getPropertyValuesByName(
+                    EgressEventTargetProperty.values(), egressEvent)
+            events.addAll(propertyValues.map {
+                baseEvent.copy(
+                        targetPropertyMaybe = it.key,
+                        newValueMaybe = it.value)
+            })
+        }
+        if (comment != null) {
```

Comment:
Kotlin trick: If we had multiple threads, then in some cases `comment` could change value in the block when using `if` but not when using the operator `?.let()`.

So I believe you can do `comment?.let {
            events.add(baseEvent.copy(
                    targetPropertyMaybe = EgressEventCommentTargetProperty.COMMENT.propertyName,
                    newValueMaybe = comment
            ))
        }
`

https://discuss.kotlinlang.org/t/let-vs-if-not-null/3542/6